treegp/two_pcf.py

```diff
@@ -0,0 +1,410 @@
+    be discribed as a Radial Basis Function such as a Gaussian kernel or a
+    von Karman kernel.
+
+    :param kernel:    sklearn.gaussian_process.kernels
+    :param x:         x coordinates of the 2D two point correlation function.
+    :param y:         y coordinates of the 2D two point correlation function.
+    :param flat_data: flatten 2D two point correlation function.
+    :param W:         Inverse of the covariance matrix got from Bootstrap.
+    :param mask:      Mask symetric area for Radial basi Function.
+    """
+    def __init__(self, kernel, flat_data, x, y, W, mask=None):
+
+        if mask is None:
+            self.mask = np.array([True]*len(x))
+        else:
+            self.mask = mask
+
+        self.kernel_class = get_kernel_class(kernel)
+        self.flat_data = flat_data
+        self.x = x
+        self.y = y
+        self.coord = np.array([x, y]).T
+        self.W = W
+        self.N = int(np.sqrt(len(self.x)))
+
+    def _model_skl(self, sigma, corr_length, g1, g2):
+        """
+        Build analytical two point correlation function
+        from sklearn kernel.
+
+        :param sigma:         Standard deviation of the gaussian random field.
+        :param corr_length:   Correlation lenght of the kernel.
+        :param g1, g2:        Shear applied to isotropic kernel.
+        """
+        if abs(g1)>1 or abs(g2)>1:
+            return None
+        else:
+            L = get_correlation_length_matrix(corr_length, g1, g2)
+            invLam = np.linalg.inv(L)
+            kernel_used = sigma**2 * self.kernel_class(invLam=invLam)
+            pcf = kernel_used.__call__(self.coord,Y=np.zeros_like(self.coord))[:,0]
+            self.kernel_fit = kernel_used
+            return pcf
+
+    def chi2(self, param):
+        """
+        Chi2 function that iminuit will minimize.
+
+        Note: only non linear parameters are searched during
+        the minimization. As linear parameters have a unique
+        solution for a given set of non-linear parameters,
+        am solving the equation to get the analytical solution
+        for the linear parameter (alpha).
+
+        :param param: list of non-linear parameters to search.
+                      (correlation lenght, e1, and e2).
+        """
+        model = self._model_skl(1., param[0],
+                                param[1], param[2])
+        if model is None:
+            self.chi2_value = np.inf
+        else:
+            model = model[self.mask]
+            F = np.array([model, np.ones_like(model)]).T
+            FWF = np.dot(F.T, self.W).dot(F)
+            Y = self.flat_data[self.mask].reshape((len(model), 1))
+            self.alpha = np.linalg.inv(FWF).dot(np.dot(F.T, self.W).dot(Y))
+            self.alpha[0] = abs(self.alpha[0])
+            self.residuals = self.flat_data[self.mask] - ((self.alpha[0] * model) + self.alpha[1])
+            self.chi2_value = self.residuals.dot(self.W).dot(self.residuals.reshape((len(model), 1)))
```

Comment:
It would also be helpful to have some documentation about where this comes from.

Reply:
For the specific case of an anisotropic kernel, you have linear parameters (the amplitude and a constant) in addition of non linear parameters. For a given set of non linear parameters, there is only one solution for the linear parameter and this is what is solved with the line 130. I can add that in comment if you think this can clarify this part of the code.